### PR TITLE
Add eventuals-grpc `RouteGuide` tests

### PR DESCRIPTION
**Pull request by onelxj**

### CODE REVIEW

File: test/grpc/route_guide/route_guide_eventuals_server.h [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+#include "test/grpc/route_guide/route_guide.grpc.pb.h"
+#include "test/grpc/route_guide/route_guide_utilities.h"
+
+using eventuals::Synchronizable;
+using eventuals::grpc::ServerReader;
+using routeguide::Feature;
+using routeguide::Point;
+using routeguide::RouteNote;
+using routeguide::eventuals::RouteGuide;
+
```
Comment (ArthurBandaryk) [inline test/grpc/route_guide/route_guide_eventuals_server.h:35]:
```

////////////////////////////////////////////////////////////////////////

```

File: test/grpc/route_guide/route_guide_eventuals_server.h [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  //                // return RequestType_();
+  //                // return RouteSummary();
+  //              });
+  //        });
+  // }
+
+ private:
+  std::vector<Feature> feature_list_;
+  std::vector<RouteNote> received_notes_;
+};
```
Comment (ArthurBandaryk) [inline test/grpc/route_guide/route_guide_eventuals_server.h:112]:
```
};

////////////////////////////////////////////////////////////////////////

```

File: test/grpc/route_guide/route_guide_eventuals_server.h [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  //                  });
+  //         }))
+  //     | Map([](auto) {
+  //          return eventuals::Catch()
+  //              .raised<std::exception>([](std::exception&& e) {
+  //                // return RequestType_();
+  //                // return RouteSummary();
+  //              });
+  //        });
+  // }
```
Comment (ArthurBandaryk) [inline test/grpc/route_guide/route_guide_eventuals_server.h:107]: Are you going to reimplement this later? If no - just delete dead code please!

File: test/grpc/route_guide/route_guide_utilities.h [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+#include <string>
+#include <vector>
+
+#include "test/grpc/route_guide/route_guide.grpc.pb.h"
+
+using routeguide::Feature;
+using routeguide::Point;
+using routeguide::RouteNote;
+
```
Comment (ArthurBandaryk) [inline test/grpc/route_guide/route_guide_utilities.h:29]:
```

////////////////////////////////////////////////////////////////////////

```

File: test/grpc/route_guide/route_guide_utilities.h [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+inline Point MakePoint(long latitude, long longitude) {
+  Point p;
+  p.set_latitude(latitude);
+  p.set_longitude(longitude);
+  return p;
+}
+
```
Comment (ArthurBandaryk) [inline test/grpc/route_guide/route_guide_utilities.h:36]:
```

////////////////////////////////////////////////////////////////////////

```

File: test/grpc/route_guide/route_guide_utilities.h [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+inline Feature MakeFeature(
+    const std::string& name,
+    long latitude,
+    long longitude) {
+  Feature f;
+  f.set_name(name);
+  f.mutable_location()->CopyFrom(MakePoint(latitude, longitude));
+  return f;
+}
+
```
Comment (ArthurBandaryk) [inline test/grpc/route_guide/route_guide_utilities.h:46]:
```

////////////////////////////////////////////////////////////////////////

```

File: test/grpc/route_guide/route_guide_utilities.h [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+inline RouteNote MakeRouteNote(
+    const std::string& message,
+    long latitude,
+    long longitude) {
+  RouteNote n;
+  n.set_message(message);
+  n.mutable_location()->CopyFrom(MakePoint(latitude, longitude));
+  return n;
+}
+
```
Comment (ArthurBandaryk) [inline test/grpc/route_guide/route_guide_utilities.h:56]:
```

////////////////////////////////////////////////////////////////////////

```

File: test/grpc/route_guide/route_guide_utilities.h [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+inline float ConvertToRadians(float num) {
+  return num * 3.1415926 / 180;
+}
+
```
Comment (ArthurBandaryk) [inline test/grpc/route_guide/route_guide_utilities.h:60]:
```

////////////////////////////////////////////////////////////////////////

```

File: test/grpc/route_guide/route_guide_utilities.h [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    const Point& point,
+    const std::vector<Feature>& feature_list) {
+  for (const Feature& f : feature_list) {
+    if (f.location().latitude() == point.latitude()
+        && f.location().longitude() == point.longitude()) {
+      return f.name();
+    }
+  }
+  return "";
+}
```
Comment (ArthurBandaryk) [inline test/grpc/route_guide/route_guide_utilities.h:93]:
```
}

////////////////////////////////////////////////////////////////////////

```

**Review by ArthurBandaryk — CHANGES_REQUESTED** [review]
Great work, Zakhar! Please address changes requested and let Ben rereview it!

File: test/grpc/route_guide/helper.cc [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+  std::string db_content(db);
+  db_content.erase(
+      std::remove_if(db_content.begin(), db_content.end(), isspace),
+      db_content.end());
+
+  Parser parser(db_content);
+  Feature feature;
+  while (!parser.Finished()) {
+    feature_list->push_back(Feature());
+    ASSERT_TRUE(parser.TryParseOne(&feature_list->back()));
```
[the remark this answers is not in the document]
Comment (onelxj) [reply test/grpc/route_guide/helper.cc:137]: Same

File: test/grpc/route_guide/helper.cc [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+#include <vector>
+
+#include "test/grpc/route_guide/route_guide.grpc.pb.h"
+
+namespace routeguide {
+
+void GetDbFileContent(const std::string& db_path, std::string& db) {
+  std::ifstream db_file(db_path);
+
+  ASSERT_TRUE(db_file.is_open());
```
[the remark this answers is not in the document]
Comment (ArthurBandaryk) [reply test/grpc/route_guide/helper.cc:35]: But `CHECK` macros will abort the application. So I do not see the reason why we can not use them!

File: test/grpc/route_guide/helper.cc [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+    return false;
+  }
+
+  bool Match(const std::string& prefix) {
+    bool eq = db_.substr(current_, prefix.size()) == prefix;
+    current_ += prefix.size();
+    return eq;
+  }
+
+  void ReadLong(long* l) {
```
[the remark this answers is not in the document]
Comment (onelxj) [reply test/grpc/route_guide/helper.cc:106]: it just returns long value from db, generic name

File: test/grpc/route_guide/route-guide-eventuals-server.h [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+  //   ... foreach: notes {
+  //     note ->
+  //       server is:
+  //         Idle { received-notes } ->
+  //           set: server is: Chatting
+  //           ... foreach: received-notes {
+  //             create and send new note
+  //           }
+  //         _ -> skip: Idle
+  //   }
```
Comment (benh) [inline test/grpc/route_guide/route-guide-eventuals-server.h:114]: You can delete this.

File: test/grpc/route_guide/route-guide-utilities.h [hunk 14]
```diff
@@ -0,0 +1,99 @@
+#pragma once
```
Comment (benh) [inline test/grpc/route_guide/route-guide-utilities.h:1]: I want to copy what is in https://github.com/grpc/grpc/tree/master/examples/cpp/route_guide as closely as possible (everything but eventuals). Please update accordingly and remove this utilities file, thanks!

File: test/grpc/route_guide/route_guide.proto [hunk 15]
```diff
@@ -0,0 +1,108 @@
+// Copyright 2015 gRPC authors.
```
Comment (benh) [inline test/grpc/route_guide/route_guide.proto:1]: Rather than copying this file can we just reference it from `@com_github_grpc_grpc` similar to what we do with `helloworld.proto`?

File: test/grpc/route_guide/BUILD.bazel [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+cc_library(
+    name = "route-guide-eventuals",
+    srcs = ["route-guide-eventuals-generated"],
+    deps = [
+        ":route-guide-grpc",
+        "//eventuals/grpc",
+    ],
+)
+
+cc_library(
```
Comment (benh) [inline test/grpc/route_guide/BUILD.bazel:41]: Can we instead just use https://github.com/grpc/grpc/blob/master/examples/cpp/route_guide/BUILD#L18?

File: test/grpc/route_guide/route-guide-eventuals-client.h [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+
+using eventuals::grpc::Client;
+using eventuals::grpc::CompletionPool;
+using routeguide::Feature;
+using routeguide::Point;
+using stout::Borrowable;
+
+////////////////////////////////////////////////////////////////////////
+
+class RouteGuideBaseClient {
```
Comment (benh) [inline test/grpc/route_guide/route-guide-eventuals-client.h:21]: Was compiling really slow and that's why you did this? I want to be as close as possible to the reference implementation so let's put the entire client in a single file as long as it is not really slow to compile. And if it is, can we declare each of the functions as returning `auto` and then only define the functions in `.cc` files where we also have the tests?

File: test/grpc/route_guide/route-chat.cc [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+
+  ASSERT_TRUE(server);
+
+  Borrowable<CompletionPool> pool;
+
+  RouteGuideRouteChatClient guide(
+      server_address,
+      grpc::InsecureChannelCredentials(),
+      pool.Borrow(),
+      db);
```
Comment (benh) [inline test/grpc/route_guide/route-chat.cc:127]: Let's create a test fixture that does all of this for us and all we need to do is call the function, e.g., `RouteChat()`.

**Review by benh — CHANGES_REQUESTED** [review]

**Review by ArthurBandaryk — APPROVED** [review]

File: test/grpc/route_guide/route-guide-eventuals-client.h [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+
+ public:
+  RouteGuideTest()
+    : guide(
+        server_address_,
+        ::grpc::InsecureChannelCredentials(),
+        pool_.Borrow(),
+        "") {}
+
+  RouteGuideClient guide;
```
Comment (benh) [inline test/grpc/route_guide/route-guide-eventuals-client.h:111]: Let's move public things to the top please!

File: test/grpc/route_guide/route-guide-eventuals-client.h [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+////////////////////////////////////////////////////////////////////////
+
+class RouteGuideTest : public ::testing::Test {
+ public:
+  const std::string db_path_ = "test/grpc/route_guide/route_guide_db.json";
+  const std::string server_address_ = "localhost:8888";
+
+  RouteGuideImpl service_;
+  std::unique_ptr<eventuals::grpc::Server> server_;
+  Borrowable<CompletionPool> pool_;
```
Comment (benh) [inline test/grpc/route_guide/route-guide-eventuals-client.h:79]: private!

File: test/grpc/route_guide/route-guide-eventuals-client.h [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+    builder.RegisterService(&service_);
+
+    auto build = builder.BuildAndStart();
+
+    ASSERT_TRUE(build.status.ok()) << build.status;
+    server_.swap(build.server);
+    ASSERT_TRUE(server_);
+
+    guide.SetDb(db);
+  }
```
Comment (benh) [inline test/grpc/route_guide/route-guide-eventuals-client.h:101]: Let's put the definition of this in a `.cc` file that also includes the server implementation.

**Review by benh — CHANGES_REQUESTED** [review]

File: protoc-gen-eventuals/templates/eventuals-client.h.j2 [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+using grpc::Status;
+
+using eventuals::grpc::Stream;
+using eventuals::grpc::Client;
+using eventuals::grpc::CompletionPool;
+using stout::Borrowable;
+using namespace {{ namespaces | join('::') }}::eventuals;
+
+{% for service in services %}
+class {{ service.name }}Client {
```
Comment (benh) [inline protoc-gen-eventuals/templates/eventuals-client.h.j2:34]: As we talked about before, we wanted to put this inside of the existing service, so for `RouteGuide` we'd have `RouteGuide::Client`. Did you run into something that made this not be a good option?

File: protoc-gen-eventuals/templates/eventuals-client.h.j2 [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+public:
+
+  {{ service.name }}Client(
+      const std::string& target,
+      const std::shared_ptr<::grpc::ChannelCredentials>& credentials,
+      stout::borrowed_ref<CompletionPool> pool) : client_(target, credentials, std::move(pool)) {}
```
Comment (benh) [inline protoc-gen-eventuals/templates/eventuals-client.h.j2:40]:
```suggestion
      stout::borrowed_ref<CompletionPool> pool) : client_(target, credentials, std::move(pool)) = default;
```

File: protoc-gen-eventuals/templates/eventuals-client.h.j2 [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+using eventuals::Let;
+using eventuals::Then;
+
+using grpc::Status;
+
+using eventuals::grpc::Stream;
+using eventuals::grpc::Client;
+using eventuals::grpc::CompletionPool;
+using stout::Borrowable;
+using namespace {{ namespaces | join('::') }}::eventuals;
```
Comment (benh) [inline protoc-gen-eventuals/templates/eventuals-client.h.j2:31]: We don't want to pollute namespaces by doing this in a header.

File: protoc-gen-eventuals/templates/eventuals-client.h.j2 [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+      >> Then(Let([input = std::move(input)](auto& call) {
+           return call.Writer().WriteLast(input)
+               >> call.Reader().Read()
+               >> Head()
+               >> Finally(Let([&](auto& output) {
+                    return call.Finish()
+                        >> Then([&](Status&& status) {
+                             CHECK(status.ok()) << status.error_code()
+                                                << ": "
+                                                << status.error_message();
```
Comment (benh) [inline protoc-gen-eventuals/templates/eventuals-client.h.j2:69]: This crashes the program if a call fails. Propagate an error instead.

File: protoc-gen-eventuals/templates/eventuals-client.h.j2 [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+                             return std::move(output);
+                           });
+                  }));
+         }));
+  }
+{%- else %}
+  auto {{ method.name }}() {
+    return client_.Call<{{ namespaces | join('::') }}::{{ service.name }}, {{ input_type }}, {{ output_type }}>("{{ method.name }}");
```
Comment (benh) [inline protoc-gen-eventuals/templates/eventuals-client.h.j2:77]: Line wrap please.

File: protoc-gen-eventuals/templates/eventuals-client.h.j2 [hunk 27]
```diff
@@ -0,0 +1,85 @@
+#pragma once
```
Comment (benh) [inline protoc-gen-eventuals/templates/eventuals-client.h.j2:1]: Split out the client changes into it's own PR please.

File: test/grpc/route_guide/route-guide-eventuals-client.h [hunk 28]
```diff
@@ -0,0 +1,91 @@
...
+
+#include <string>
+#include <vector>
+
+#include "eventuals/do-all.h"
+#include "eventuals/foreach.h"
+#include "eventuals/grpc/client.h"
+#include "eventuals/range.h"
+#include "test/grpc/route_guide/route-guide-eventuals-server.h"
+#include "test/grpc/route_guide/route_guide_generated/route_guide.client.eventuals.h"
```
Comment (benh) [inline test/grpc/route_guide/route-guide-eventuals-client.h:13]:
```suggestion
#include "test/grpc/route_guide/route_guide.client.eventuals.h"
```

File: test/grpc/route_guide/route-guide-eventuals-server.h [hunk 29]
```diff
@@ -0,0 +1,207 @@
...
+#include "eventuals/filter.h"
+#include "eventuals/flat-map.h"
+#include "eventuals/iterate.h"
+#include "eventuals/let.h"
+#include "eventuals/loop.h"
+#include "eventuals/map.h"
+#include "eventuals/then.h"
+#include "test/grpc/route_guide/helper.h"
+#include "test/grpc/route_guide/route_guide.grpc.pb.h"
+#include "test/grpc/route_guide/route_guide_generated/route_guide.eventuals.h"
```
Comment (benh) [inline test/grpc/route_guide/route-guide-eventuals-server.h:16]:
```suggestion
#include "test/grpc/route_guide/route_guide.eventuals.h"
```

File: test/grpc/route_guide/server-client-implementation.cc [hunk 30]
```diff
@@ -0,0 +1,62 @@
+#include <grpcpp/channel.h>
+
+#include "route-guide-eventuals-client.h"
+#include "route-guide-eventuals-server.h"
+#include "test/grpc/route_guide/helper.h"
+#include "test/grpc/route_guide/route_guide.grpc.pb.h"
+#include "test/grpc/route_guide/route_guide_generated/route_guide.eventuals.h"
```
Comment (benh) [inline test/grpc/route_guide/server-client-implementation.cc:7]:
```suggestion
#include "test/grpc/route_guide/route_guide.eventuals.h"
```

File: protoc-gen-eventuals/templates/eventuals-client.h.j2 [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+{%- set output_type -%}
+    {%- if method.server_streaming -%}
+        Stream<{{ method.output_type.split('.') | join('::') }}>
+    {%- else -%}
+        {{ method.output_type.split('.') | join('::') }}
+    {%- endif -%}
+{%- endset %}
+{%- if not method.server_streaming and not method.client_streaming %}
+  auto {{ method.name }}({{ method.input_type.split('.') | join('::') }}&& input) {
+    return client_.Call<{{ namespaces | join('::') }}::{{ service.name }}, {{ input_type }}, {{ output_type }}>("{{ method.name }}")
```
Comment (benh) [inline protoc-gen-eventuals/templates/eventuals-client.h.j2:59]: Wrap please.

File: bazel/copts.bzl [hunk 32]
```diff
@@ -13,11 +13,15 @@ def copts():
             # We do not want to treat `syscall` warning as an error.
             # Future update of glog lib will fix this warning.
             "-Wno-error=deprecated-declarations",
+            "-std=c++17",
```
Comment (benh) [inline bazel/copts.bzl:16]: Why is this necessary?

**Review by benh — CHANGES_REQUESTED** [review]

File: protoc-gen-eventuals/templates/eventuals-client.h.j2 (hunk 23 above)
Comment (onelxj) [reply protoc-gen-eventuals/templates/eventuals-client.h.j2:40]: this can't be defaulted, since we call non default `client_` constructor
